packages/flutter_tools/lib/src/cache.dart

```diff
@@ -254,51 +254,53 @@ class FlutterEngine {
 
   List<String> _getPackageDirs() => const <String>[kSkyEngine];
 
-  List<String> _getEngineDirs() {
-    List<String> dirs = <String>[
-      'android-arm',
-      'android-arm-profile',
-      'android-arm-release',
-      'android-x64',
-      'android-x86',
-    ];
+  // Return a list of (cache directory path, download URL path) tuples.
+  List<List<String>> _getBinaryDirs() {
+    List<List<String>> binaryDirs = <List<String>>[];
 
     if (cache.includeAllPlatforms)
-      dirs.addAll(<String>['ios', 'ios-profile', 'ios-release', 'linux-x64']);
-    else if (Platform.isMacOS)
-      dirs.addAll(<String>['ios', 'ios-profile', 'ios-release']);
+      binaryDirs
+        ..addAll(_osxBinaryDirs)
+        ..addAll(_linuxBinaryDirs)
+        ..addAll(_androidBinaryDirs)
+        ..addAll(_iosBinaryDirs);
     else if (Platform.isLinux)
-      dirs.add('linux-x64');
-
-    return dirs;
-  }
-
-  // Return a list of (cache directory path, download URL path) tuples.
-  List<List<String>> _getToolsDirs() {
-    if (cache.includeAllPlatforms)
-      return <List<String>>[]
-        ..addAll(_osxToolsDirs)
-        ..addAll(_linuxToolsDirs);
+      binaryDirs
+        ..addAll(_linuxBinaryDirs)
+        ..addAll(_androidBinaryDirs);
     else if (Platform.isMacOS)
-      return _osxToolsDirs;
-    else if (Platform.isLinux)
-      return _linuxToolsDirs;
-    else
-      return <List<String>>[];
+      binaryDirs
+        ..addAll(_osxBinaryDirs)
+        ..addAll(_androidBinaryDirs)
+        ..addAll(_iosBinaryDirs);
+
+    return binaryDirs;
   }
 
-  List<List<String>> get _osxToolsDirs => <List<String>>[
+  List<List<String>> get _osxBinaryDirs => <List<String>>[
```

Comment:
Since the cache directory basename is repeated in the artifact portion of every one of these entries, we should strip it from the artifact portion and just update `download()` accordingly.